server/src/main/java/org/elasticsearch/common/time/Iso8601Parser.java

```diff
@@ -0,0 +1,486 @@
+     */
+    Iso8601Parser(Set<ChronoField> mandatoryFields, boolean optionalTime, Map<ChronoField, Integer> defaults) {
+        checkChronoFields(mandatoryFields, VALID_MANDATORY_FIELDS);
+        checkChronoFields(defaults.keySet(), VALID_DEFAULT_FIELDS);
+
+        this.mandatoryFields = EnumSet.of(ChronoField.YEAR); // year is always mandatory
+        this.mandatoryFields.addAll(mandatoryFields);
+        this.optionalTime = optionalTime;
+        this.defaults = defaults.isEmpty() ? Map.of() : new EnumMap<>(defaults);
+    }
+
+    private static void checkChronoFields(Set<ChronoField> fields, Set<ChronoField> validFields) {
+        if (fields.isEmpty()) return;   // nothing to check
+
+        fields = EnumSet.copyOf(fields);
+        fields.removeAll(validFields);
+        if (fields.isEmpty() == false) {
+            throw new IllegalArgumentException("Invalid chrono fields specified " + fields);
+        }
+    }
+
+    boolean optionalTime() {
+        return optionalTime;
+    }
+
+    Set<ChronoField> mandatoryFields() {
+        return mandatoryFields;
+    }
+
+    private boolean isOptional(ChronoField field) {
+        return mandatoryFields.contains(field) == false;
+    }
+
+    private Integer defaultZero(ChronoField field) {
+        return defaults.getOrDefault(field, 0);
+    }
+
+    Result tryParse(CharSequence str, @Nullable ZoneId defaultTimezone) {
+        if (str.charAt(0) == '-') {
+            // the year is negative. This is most unusual.
+            // Instead of always adding offsets and dynamically calculating position in the main parser code below,
+            // just in case it starts with a -, just parse the substring, then adjust the output appropriately
+            Result result = parse(new CharSubSequence(str, 1, str.length()), defaultTimezone);
+
+            if (result.errorIndex() >= 0) {
+                return Result.error(result.errorIndex() + 1);
+            } else {
+                DateTime dt = result.result();
+                return new Result(
+                    new DateTime(
+                        -dt.years(),
+                        dt.months(),
+                        dt.days(),
+                        dt.hours(),
+                        dt.minutes(),
+                        dt.seconds(),
+                        dt.nanos(),
+                        dt.zoneId(),
+                        dt.offset()
+                    )
+                );
+            }
+        } else {
+            return parse(str, defaultTimezone);
+        }
+    }
+
+    /**
+     * Index {@code i} is the multiplicand to get the number of nanos from the fractional second with {@code i=9-d} digits.
+     */
+    private static final int[] NANO_MULTIPLICANDS = new int[] { 1, 10, 100, 1_000, 10_000, 100_000, 1_000_000, 10_000_000, 100_000_000 };
+
+    private Result parse(CharSequence str, @Nullable ZoneId defaultTimezone) {
```

Comment:
this is a crucial method of the whole parsing and it is not trivial. Can we have some javadoc for this method too?

Reply:
Javadocs added